FEXCore/Source/Interface/IR/Passes/x87StackOptimizationPass.cpp

```diff
@@ -0,0 +1,1148 @@
+  Exp = IREmit->_Sub(OpSize::i64Bit, Exp, GetConstant(1023));
+  Exp = IREmit->_Float_FromGPR_S(8, 8, Exp);
+  Ref Sig = IREmit->_And(OpSize::i64Bit, Gpr, GetConstant(0x800fffffffffffffLL));
+  Sig = IREmit->_Or(OpSize::i64Bit, Sig, GetConstant(0x3ff0000000000000LL));
+  Sig = IREmit->_VCastFromGPR(8, 8, Sig);
+
+  return std::tuple {Exp, Sig};
+}
+
+void X87StackOptimization::Run(IREmitter* Emit) {
+  FEXCORE_PROFILE_SCOPED("PassManager::x87StackOpt");
+
+  auto CurrentIR = Emit->ViewIR();
+  auto* HeaderOp = CurrentIR.GetHeader();
+  LOGMAN_THROW_AA_FMT(HeaderOp->Header.Op == OP_IRHEADER, "First op wasn't IRHeader");
+
+  if (!HeaderOp->HasX87) {
+    // If there is no x87 in this, just early exit.
+    return;
+  }
+
+  // Initialize IREmit member
+  IREmit = Emit;
+
+  // Run optimization proper
+  for (auto [BlockNode, BlockHeader] : CurrentIR.GetBlocks()) {
+    auto BlockIROp = BlockHeader->CW<FEXCore::IR::IROp_CodeBlock>();
+    // Each time we deal with a new block we need to start over.
+    // The optimization should run per-block
+    Reset();
+
+    for (auto [CodeNode, IROp] : CurrentIR.GetCode(BlockNode)) {
+      if (!LoweredX87(IROp->Op)) {
+        continue;
+      }
+      IREmit->SetWriteCursor(CodeNode);
+      switch (IROp->Op) {
+      case OP_F80ADDSTACK: {
+        const auto* Op = IROp->C<IROp_F80AddStack>();
+        HandleBinopStack(OP_VFADD, true, OP_F80ADD, Op->SrcStack1, Op->SrcStack1, Op->SrcStack2);
+        IREmit->Remove(CodeNode);
+        break;
+      }
+
+      case OP_F80SUBSTACK: {
+        const auto* Op = IROp->C<IROp_F80SubStack>();
+        HandleBinopStack(OP_VFSUB, true, OP_F80SUB, Op->DstStack, Op->SrcStack1, Op->SrcStack2);
+        IREmit->Remove(CodeNode);
+        break;
+      }
+
+      case OP_F80MULSTACK: {
+        const auto* Op = IROp->C<IROp_F80MulStack>();
+        HandleBinopStack(OP_VFMUL, true, OP_F80MUL, Op->SrcStack1, Op->SrcStack1, Op->SrcStack2);
+        IREmit->Remove(CodeNode);
+        break;
+      }
+
+      case OP_F80DIVSTACK: {
+        const auto* Op = IROp->C<IROp_F80DivStack>();
+        HandleBinopStack(OP_VFDIV, true, OP_F80DIV, Op->DstStack, Op->SrcStack1, Op->SrcStack2);
+        IREmit->Remove(CodeNode);
+        break;
+      }
+
+      case OP_F80FPREMSTACK: {
+        HandleBinopStack(OP_F64FPREM, false, OP_F80FPREM, 0, 0, 1);
+        IREmit->Remove(CodeNode);
+        break;
+      }
+
+      case OP_F80FPREM1STACK: {
+        HandleBinopStack(OP_F64FPREM1, false, OP_F80FPREM1, 0, 0, 1);
+        IREmit->Remove(CodeNode);
+        break;
+      }
+
+      case OP_F80SCALESTACK: {
+        HandleBinopStack(OP_F64SCALE, false, OP_F80SCALE, 0, 0, 1);
+        IREmit->Remove(CodeNode);
+        break;
+      }
+
+      case OP_F80FYL2XSTACK: {
+        HandleBinopStack(OP_F64FYL2X, false, OP_F80FYL2X, 1, 0, 1);
+        if (SlowPath) {
+          UpdateTopForPop_Slow();
+        } else {
+          StackData.pop();
+        }
```

Comment:
This should be a helper

Reply:
I think we only have 3 of these. Does it really make sense here?

Reply:
Sometimes helpers are useful even if they're only called once for the sake of expressing the intention clearly and separating the "manage the stack" logic from the "implement all the instructions" 🤷

Reply:
sgtm - added helper.